### DISCUSSION TITLE
Custom error handler

### DISCUSSION THREAD

**Discussion by moon-stripe** [verdaccio/verdaccio#1287] [Ideas]
**Is your feature request related to a problem? Please describe.**
Similar to #1286, we use [sentry](https://sentry.io/welcome/) for error reporting, but there don't appear to be any hooks into verdaccio to add custom error reporting middleware.

**Describe the solution you'd like**
To use [sentry's express integration](https://docs.sentry.io/platforms/node/express/), you have to add a middleware before other middlewares, and an error-handling middleware before other error-handling middlewares. These seem like common enough use-cases that you could extend the middleware plugin api to something like:
```javascript
const Sentry = require('@sentry/node');
module.exports => config => {
  Sentry.init({ dsn: config.dsn });

  return {
    before_middleware: (app) => {
      /* middleware called first */
      app.use(Sentry.Handlers.requestHandler());
    },
    register_middlewares: (app) => { /* same as current middleware api */ },
    error_middleware: (app) => {
      /* error middleware called before other error middleware */
      /* could be any custom error handler */
      app.use(Sentry.Handlers.errorHandler());
    },
  };
}
```

**Describe alternatives you've considered**
* Even just an `error_middleware` api that gets called after all other routes and middleware would allow for [manual handling of errors](https://docs.sentry.io/error-reporting/capturing/?platform=node):

```javascriptconst Sentry = require('@sentry/node');
module.exports => config => {
  Sentry.init({ dsn: config.dsn });

  return {
    error_middleware: (app) => {
      /* error middleware called before other error middleware */
      /* could be any custom error handler */
      app.use((err, req, res, next) => {
        Sentry.captureException(err);
        next(err);
      });
    },
  };
}
```
* Ship with sentry support that just does all of the above, by adding a direct dependency to `@sentry/node` or an api to pass in a sentry instance. The above middleware solution seems more flexible and easier to maintain though, and I imagine would work well for most if not all error-reporting tools.

**Additional context**
Thanks for verdaccio! We've had a great experience with it so far!

**Comment by juanpicado** [comment]
hi @moon-stripe I think your request makes total sense. So, I'll set a label for discussion and future development. Thanks for your detailed input.

:-) and thanks for using verdaccio, happy you are glad.
